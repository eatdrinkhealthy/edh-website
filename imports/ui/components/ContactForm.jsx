import React from "react";
import PropTypes from "prop-types";
import { Field, reduxForm } from "redux-form";

const ValidationError = ({ children, visible }) => {
  const classes = `form-error ${visible ? "is-visible" : ""}`;

  return <span className={classes}>{children}</span>;
};

ValidationError.propTypes = {
  children: PropTypes.node,
  visible: PropTypes.bool,
};

ValidationError.defaultProps = {
  children: "",
  visible: false,
};

/* eslint-disable jsx-a11y/label-has-for */

const ContactFormComponent = () => (
  <div>
    <h2>Contact Us</h2>
    <form onSubmit={() => alert("submitted!")}>
      <div>
        <label htmlFor="contactName">Name </label>
        <Field name="contactName" component="input" type="text" autoFocus />
        <ValidationError visible>validation error.</ValidationError>
      </div>

      <div>
        <label htmlFor="contactEmail">Email </label>
        <Field name="contactEmail" component="input" type="email" />
        <ValidationError>validation error.</ValidationError>
      </div>

      <div>
        <label htmlFor="contactMessage">Message </label>
        <Field name="contactMessage" component="textarea" rows={10} cols={50} />
        <ValidationError>validation error.</ValidationError>
      </div>

      <button id="contactSubmit" type="submit" className="button button--submit">
        Send
      </button>
    </form>
  </div>
);

const ContactForm = reduxForm({ form: "contactForm" })(ContactFormComponent);

export default ContactForm;
