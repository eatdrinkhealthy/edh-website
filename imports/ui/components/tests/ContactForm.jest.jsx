// @flow
/* eslint-env jest */
/* eslint-disable func-names, prefer-arrow-callback, no-unused-expressions */
/* eslint-disable import/no-extraneous-dependencies */
import React from "react";
import { createStore } from "redux";
import { Provider } from "react-redux";
import renderer from "react-test-renderer";
import ContactForm from "../ContactForm";
import mountFormWithInputs from "../../../utils/tests/mountFormWithInputs";
import appReducer from "../../../state/reducers";

describe("<ContactForm />", () => {
  const testStore = createStore(appReducer);

  it("matches render snapshot - with no form validation errors", () => {
    // render form with no input values to display validation errors
    const tree = renderer
      .create(
        <Provider store={testStore}>
          <ContactForm onSubmit={() => {}} />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should set Name 'too short' error, when Name less than min length", () => {
    const props = {
      onSubmit: jest.fn(),
    };

    const wrapper = mountFormWithInputs(
      <ContactForm {...props} />,
      {
        contactName: "a",
      },
      testStore,
    );

    wrapper.find("#contactName").simulate("blur"); // blur triggers field validation

    expect(wrapper.find("#contactNameError").text()).toBe(
      "Must have a least 2 characters.",
    );
  });

  it("should set invalid email error, when email format is invalid", () => {
    const props = {
      onSubmit: jest.fn(),
    };

    const wrapper = mountFormWithInputs(
      <ContactForm {...props} />,
      {
        contactEmail: "user12",
      },
      testStore,
    );

    wrapper.find("#contactEmail").simulate("blur"); // blur triggers field validation

    expect(wrapper.find("#contactEmailError").text()).toBe("Invalid email address.");
  });

  it("should set message 'too short' error, when message is too short", () => {
    const props = {
      onSubmit: jest.fn(),
    };

    const wrapper = mountFormWithInputs(
      <ContactForm {...props} />,
      {
        contactMessage: "1",
      },
      testStore,
    );

    wrapper.find("#contactMessage").simulate("blur"); // blur triggers field validation

    expect(wrapper.find("#contactMessageError").text()).toBe(
      "Must have a least 2 characters.",
    );
  });

  it("should set message 'too long' error, when message is too long", () => {
    const props = {
      onSubmit: jest.fn(),
    };

    const wrapper = mountFormWithInputs(
      <ContactForm {...props} />,
      {
        contactMessage:
          "12345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890" +
          "12345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890" +
          "12345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890" +
          "12345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890" +
          "12345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890",
      },
      testStore,
    );

    wrapper.find("#contactMessage").simulate("blur"); // blur triggers field validation

    expect(wrapper.find("#contactMessageError").text()).toBe(
      "Must not have more than 500 characters.",
    );
  });

  it("should call onSubmit, when submitting and all form fields are valid", () => {
    const props = {
      onSubmit: jest.fn(),
    };

    const wrapper = mountFormWithInputs(
      <ContactForm {...props} />,
      {
        contactName: "user12",
        contactEmail: "user12@test.com",
        contactMessage: "a message to send.",
      },
      testStore,
    );

    wrapper.find("#contactSubmit").simulate("submit");
    expect(props.onSubmit).toHaveBeenCalledWith(
      {
        contactName: "user12",
        contactEmail: "user12@test.com",
        contactMessage: "a message to send.",
      },
      expect.anything(),
      expect.anything(),
    );
  });

  it("should clear name, email, and message on successful submit", () => {
    const props = {
      onSubmit: jest.fn(),
    };

    const wrapper = mountFormWithInputs(
      <ContactForm {...props} />,
      {
        contactName: "user12",
        contactEmail: "user12@test.com",
        contactMessage: "a message to send.",
      },
      testStore,
    );
    const contactNameNode = wrapper.find("#contactName");
    const contactEmailNode = wrapper.find("#contactEmail");
    const contactMessageNode = wrapper.find("#contactMessage");

    // give focus to contactMessageNode input (like a user would do before submit)
    contactEmailNode.prop("onFocus")();
    wrapper.find("#contactSubmit").simulate("submit");

    expect(contactNameNode.props().value).toEqual("");
    expect(contactEmailNode.props().value).toEqual("");
    expect(contactMessageNode.props().value).toEqual("");
    // NOTE: document.querySelector called in onSubmitSuccess works in browser but not in jest
    // expect(usernameNode.get(0)).toEqual(document.activeElement);
  });

  it("should set focus to Name input on render", () => {
    const props = {
      onSubmit: jest.fn(),
    };

    mountFormWithInputs(
      <ContactForm {...props} />,
      {
        contactName: "user12",
        contactEmail: "user12@test.com",
        contactMessage: "a message to send.",
      },
      testStore,
    );

    expect(document.activeElement.id).toBe("contactName");
  });
});
