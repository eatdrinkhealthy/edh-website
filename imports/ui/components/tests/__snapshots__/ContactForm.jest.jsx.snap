// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContactForm /> matches render snapshot - with no form validation errors 1`] = `
<div>
  <h2>
    Contact Us
  </h2>
  <form
    onSubmit={[Function]}
  >
    <div>
      <label
        htmlFor="contactName"
      >
        Name
      </label>
      <input
        autoFocus={true}
        component="input"
        id="contactName"
        name="contactName"
        onBlur={[Function]}
        onChange={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        type="text"
        value=""
      />
    </div>
    <div>
      <label
        htmlFor="contactEmail"
      >
        Email
      </label>
      <input
        autoFocus={false}
        component="input"
        id="contactEmail"
        name="contactEmail"
        onBlur={[Function]}
        onChange={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        type="email"
        value=""
      />
    </div>
    <div>
      <label
        htmlFor="contactMessage"
      >
        Message
      </label>
      <textarea
        cols="10"
        id="contactMessage"
        name="contactMessage"
        onBlur={[Function]}
        onChange={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        rows="50"
        value=""
      />
    </div>
    <button
      className="button button--submit"
      id="contactSubmit"
      type="submit"
    >
      Send
    </button>
  </form>
</div>
`;
